keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md

```diff
@@ -99,105 +101,152 @@ A developer has a Windows web app that they would like to deploy in a container
+  - Create a rolebinding and assign desired service accounts to the above role:
 
 ```
-kind: RoleBinding
 apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
 metadata:
-  name: use-webserver-gmsa
+  name: use-webapp1-gmsa
   namespace: default
 subjects:
-- kind: User
-  name: jane
+- kind: ServiceAccount
+  name: webapp-sa
   apiGroup: rbac.authorization.k8s.io
 roleRef:
   kind: Role
-  name: webserver-gmsa-user
+  name: webapp1-gmsa-user
   apiGroup: rbac.authorization.k8s.io
 ```
   
-4. Application admins deploy app pods with a reference to the name of an authorized GMSA configmap (e.g. `webserver-credspec`). In the alpha and beta stage of this feature, the reference to the GMSA configmap will be set through an annotation on the pod: `pod.[alpha|beta].kubernetes.io/windows-gmsa-config-map`. In the beta and GA stage, the reference to GMSA configmap may be set through `spec.securityContext.windows.credentialSpecConfig` in the pod spec.
-5. During pod creation, a new Admission Controller, GMSAAuthorizer, will ensure the `user` (and `spec.serviceAccountName` if specified) is authorized for the `use` verb on the GMSA configmap. If the authorization check fails due to absence of RBAC roles, the pod creation fails.
-6. The Windows CRI implementation accesses the configmap with the GMSA credspec and copies the contents of entry with key `gmsa-cred-spec` to the [OCI windows.CredentialSpec](https://github.com/opencontainers/runtime-spec/blob/master/config-windows.md#credential-spec) field.
-7. The Windows OCI implementation validates the credspec and fails to start the container if its invalid or access to the GMSA from the node is denied.
-8. The container starts with the Windows Group Managed Service Account as expected, and can authenticate to the database.
-9. During any subsequent pod update, any changes to a pod's `securityContext` is blocked (as is the case today) by `ValidatePodUpdate`.
+5. Application admins will deploy app pods that require a GMSA identity along with a Service Account authorized to use the GMSAs. There will be two options regarding how to specify the GMSACredSpec for pods and containers:
+
+  - Specify the name of the desired GMSACredSpec (e.g. `webapp1-credspec`): In the alpha stage of this feature, the name of the desired GMSACredSpec can be set through an annotation on the pod: `pod.alpha.kubernetes.io/windows-gmsa-credspec-name` as well as for each container through annotations of the form: `container.windows-gmsa-credspec-name.alpha.kubernetes.io/container-name`. In the beta stage, the annotations will be promoted to a field in the securityContext of the pod: `podspec.securityContext.windows.gmsaCredSpecName` and in the securityContext of each container:  `podspec.container[i].securityContext.windows.gmsaCredSpecName`. The GMSACredSpec name for a container will override the GMSACredSpec entries for the whole pod.
```

Comment:
I don't know if there are precedents within k8s for these annotation names, but wouldn't something like `container.alpha.kubernetes.io/windows-gmsa-credspec-name/<container-name>` be more consistent with the pod-level one (by not having the "kind" `windows-gmsa-credspec-name` in the "hostname" portion)?

Reply:
I'll try with 2 `/`s first, and if not go with yours if that's okay then.

Reply:
The 2 `/`s is failing for something like `test.k8s.io/container/foo`: 
```The Deployment "nginx-deployment" is invalid: spec.template.annotations: Invalid value: "test.k8s.io/container/foo": a qualified name must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]') with an optional DNS subdomain prefix and '/' (e.g. 'example.com/MyName')```